v1

ajvb opened the pull request:
Initial version of a Lambda function that will ingest syslog-over-https, specifically targeting Heroku [Log Drains](https://devcenter.heroku.com/articles/log-drains) to start.

The AWS infrastructure pieces are:
* API Gateway with Lambda Proxy integration
* DynamoDB table with username/password combos
* Kinesis stream that records are pushed to

As well, I am waiting on adding the CloudFormation templates until I get an initial review, especially as I am unsure about using DynamoDB and whether we actually want to do that.

r? @g-k @ameihm0912

ajvb committed:
v1 pre-review

ajvb commented on utils.py:
```diff
@@ -0,0 +1,108 @@
+    return [l for l in parser.parse_rfc6587(body) if filter_log_lines(l)]
+
+
+def filter_log_lines(msg):
```
This works as an initial placeholder, but I think this will be the first piece to be reworked. Will probably want some not-hardcoded way to include filters, as there will be a lot.

ajvb committed:
Small stylistic tweaks and fixes

ajvb committed:
mistake in fixture

ajvb committed:
Fixed the FIXME comment within the syslog parsing code

ajvb commented on utils.py:
```diff
@@ -0,0 +1,112 @@
+        num_of_unicode_chars = len(
+            [c for c in syslog_msg_payload[0:msg_len] if ord(c) > 128]
+        )
+        msg_len -= num_of_unicode_chars
```
I would love some assistance on what is the best way to accomplish the above.

#### As well, here is an example:
```
In [16]: logline = '155 <190>1 2018-08-03T22:38:42.975461+00:00 host app web.1 - [GIN] 2018/08/03 - 22:
    ...: 38:42 | 200 |     170.221µs |  256.0.0.10 | GET      /static/lang-logo.png\n146 <190>1 2018-08
    ...: -03T22:38:43.087424+00:00 host app web.1 - [GIN] 2018/08/03 - 22:38:43 | 404 |       1.973µs |
    ...:   256.0.0.10 | GET      /favicon.ico\n'

In [17]: len("<190>1 2018-08-03T22:38:42.975461+00:00 host app web.1 - [GIN] 2018/08/03 - 22:38:42 | 20
    ...: 0 |     170.221µs |  256.0.0.10 | GET      /static/lang-logo.png\n")
Out[17]: 154

In [18]: logline[4:]
Out[18]: '<190>1 2018-08-03T22:38:42.975461+00:00 host app web.1 - [GIN] 2018/08/03 - 22:38:42 | 200 |     170.221µs |  256.0.0.10 | GET      /static/lang-logo.png\n146 <190>1 2018-08-03T22:38:43.087424+00:00 host app web.1 - [GIN] 2018/08/03 - 22:38:43 | 404 |       1.973µs |  256.0.0.10 | GET      /favicon.ico\n'

# The last character is the first digit in the next messages MSGLEN
In [19]: logline[4:][0:155]
Out[19]: '<190>1 2018-08-03T22:38:42.975461+00:00 host app web.1 - [GIN] 2018/08/03 - 22:38:42 | 200 |     170.221µs |  256.0.0.10 | GET      /static/lang-logo.png\n1'
```

g-k replied:
Is the MSGLEN a byte len vs str len thing? Can you used the byte len?

```
>>> len(bytes('¯\_(ツ)_/¯', 'utf8'))
13
>>> len('¯\_(ツ)_/¯')
9
```

ameihm0912 commented on Makefile:
```diff
@@ -0,0 +1,47 @@
+
+.PHONY:
+	all \
+	black \
```
I think the build related targets might be missing here

ameihm0912 commented on Makefile:
```diff
@@ -0,0 +1,47 @@
+	(cd ${VIRTUAL_ENV}/lib/python3.6/site-packages && zip -r9 ${VIRTUAL_ENV}/../pylogdrain.zip .)
+	zip -g pylogdrain.zip *.py
+
+docker_build: install build
```
Should we also depend on `clean` here since it's occurring in a volume mount?

ameihm0912 commented on auth.py:
```diff
@@ -0,0 +1,104 @@
+
+        else:
+            raise BasicAuthDecodeError
+
```
Do we need any additional checks here to validate the b64 decoder/split didn't for example return a single element or something else we didn't expect

ajvb replied:
I don't think so, but I could see some kind of error occurring from an encoding problem, just not sure if it is the responsibility of this function to handle that. In regards to returning a single element, you will get a `ValueError` exception raised if that happens.

ameihm0912 commented on auth.py:
```diff
@@ -0,0 +1,104 @@
+            return False
+
+        resp = self.dynamo_client.get_item(
+            TableName=self.dynamo_table, Key={"username": {"S": username}}
```
Yeah I'm wondering if the Dynamo table doesn't add complexity considering it's basically going to just be read only and only contain a few items -- one other option I guess too could be we just pass the entire configuration for this function in as a JSON doc in an environment variable, that just get's decrypted via sops when we deploy it

ajvb replied:
I'm good with that. I'll refactor it to grab a json doc stored in S3 that is encrypted using sops. Do you think we should stick with using bcrypt for the passwords?

ameihm0912 commented on utils.py:
```diff
@@ -0,0 +1,112 @@
+
+
+class Parser(object):
+    def __init__(self):
```
I'm not sure on this part of the code,

What would be ideal is we just simply split the octet framed individual syslog messages out, and return them as a list and send them that way without parsing them. So, basically if the block contains 3 messages, we'd end up parsing that into 3 strings and just submitting those as is; they'd be converted into a list as though they came in individually.

We can then just use hindsight's syslog decoders to do any additional parsing that is required and it will be consistent across the board with other syslog imports and normalized the same way.

ameihm0912 replied:
One other thought here too -- would the intent be that there would be a single API gateway for multiple drains? In the scenario would there be a way to tell which Heroku app sent the log message?

If not should we run a gateway/lambda per application category or something, or does the log message contain something we could use to identify where it came from?

ajvb replied:
This sounds great. I would love to remove this section of the code. The one question I have though is we had initially discussed filtering out messages within the Lambda func, should we not do this? If we do still want to filter out some messages then I imagine the func will need to parse out the contents of the syslog log line.

>If not should we run a gateway/lambda per application category or something, or does the log message contain something we could use to identify where it came from?

In looking at the logs, sometimes they do sometimes they do not. I think an easy way to accomplish this is a set of credentials per app+env and we can use the username as the identifier.

ajvb replied:
> The one question I have though is we had initially discussed filtering out messages within the Lambda func, should we not do this?

I'm not exactly sure what all we will be filtering out at this moment, but we could definitely start with filtering out based on regex rules applied to a single log line, negating the need parse it out.

ameihm0912 commented on utils.py:
```diff
@@ -0,0 +1,112 @@
+        payload["hostname"] = parsed[3]
+        payload["source"] = parsed[4]
+        payload["appname"] = parsed[5]
+        payload["message"] = parsed[6]
```
See comment above, this will require creating a new decoder to make use our of existing syslog parsers, we should just send the string as is

ameihm0912 reviewed (commented):
👍

ajvb committed:
CR fixes

g-k commented on README.md:
```diff
@@ -0,0 +1,57 @@
+
+It uses a DynamoDB table to store username/password combinations, with the password being
+bcrypt hashed. After validating that a request has a working use of credentials, it will
+send the logs to a Kinesis stream, one log line per record.
```
What are the username/passwd combos for? Any reason we're using DynamoDB in particular?

ajvb replied:
>What are the username/passwd combos for?

syslog-over-https within Heroku (and I believe other systems as well) only supports Basic Auth, so we are using that for authentication for log ingestion.

>Any reason we're using DynamoDB in particular?

Just because it's simple, but per https://github.com/mozilla-services/pylogdrain/pull/1#discussion_r208775772 I'm going to refactor it to use a JSON doc in S3 instead.

g-k reviewed (commented):
Not sure I totally understand the requirements here, but it is looking good overall.

If this is just receiving the messages, I'd prefer saving them in raw format (maybe w/ PII and creds filtered) and doing the parsing in a separate lambda or hindsight.

ajvb committed:
handle the fact that msglen is bytes len rather than str len

ajvb committed:
Refactor informed by code review

* Remove parsing the syslog message, and instead just send the message
  directly to kinesis.
* Move from using Dynamo to using a json document stored in S3

ajvb committed:
Removed unnecessary json encoding

ajvb commented:
@g-k 

>If this is just receiving the messages, I'd prefer saving them in raw format (maybe w/ PII and creds filtered) and doing the parsing in a separate lambda or hindsight.

:+1: - changed to just splitting out each log line and sending them directly to kinesis (for hindsight to parse)

>Is the MSGLEN a byte len vs str len thing? Can you used the byte len?

Yes! Thank you, that is exactly what I was looking for.

ajvb committed:
black fixes + removed dynamo key in log message

ajvb committed:
Add cloudformation template and update README

ajvb commented:
@ameihm0912 Ok, changed to using a json doc in s3 and removed the parsing of the individual syslog message. As well, added a dev cloudformation template

g-k reviewed (approved):
r+ :+1: lgtm

ameihm0912 commented on auth.py:
```diff
@@ -0,0 +1,111 @@
+    def check_header(self, basicauth_header):
+        """
+        Checks Basic Auth header value against username/password combos
+        in dynamodb.
```
not sure if we care about updating this old comment

ajvb committed:
update check_header docstring
